cgp/genome.py

```diff
@@ -267,6 +267,162 @@ def randomize(self, rng: np.random.RandomState) -> None:
 
+    def reorder(self, rng: np.random.RandomState) -> None:
+        """Reorder the genome
+
+        Shuffle node ordering of internal (hidden) nodes in genome without changing the phenotype.
+        (Goldman 2015, DOI: 10.1109/TEVC.2014.2324539)
+
+        During reordering, inactive genes, e.g., input genes of nodes with arity zero,
+        are not taken into account and can hence have invalid values after reordering.
+        These invalid values are replaced by random values
+        for the respective gene after reordering.
+
+        Parameters
+        ----------
+        rng : numpy.RandomState
+            Random number generator instance.
+
+        Returns
+        ----------
+        None
+        """
+        if (self._n_rows != 1) or (self._levels_back != self._n_columns):
+            raise ValueError(
+                "Genome reordering is only implemented for n_rows=1" " and levels_back=n_columns"
+            )
+
+        dna = self._dna.copy()
+
+        node_dependencies: Dict[int, Set[int]] = self._determine_node_dependencies()
+
+        addable_nodes: Set[int] = self._get_addable_nodes(node_dependencies)
+
+        new_node_idx: int = self._n_inputs  # First position to be placed is after inputs
+        used_node_indices: List[int] = []
+
+        while len(addable_nodes) > 0:
+
+            old_node_idx = rng.choice(list(addable_nodes))
+
+            dna = self._copy_dna_segment(dna, old_node_idx=old_node_idx, new_node_idx=new_node_idx)
+
+            for dependencies in node_dependencies.values():
+                dependencies.discard(old_node_idx)
+
+            addable_nodes, used_node_indices = self._update_addable_nodes(
+                node_dependencies, used_node_indices, old_node_idx=old_node_idx
+            )
+            new_node_idx += 1
+
+        self._update_input_genes(dna, used_node_indices)
+        self._replace_invalid_input_gene(dna, rng)
+
+        self.dna = dna
+
+    def _copy_dna_segment(self, dna: List[int], old_node_idx: int, new_node_idx: int) -> List[int]:
+        """ Copy a nodes dna segment from its old node location to a new location. """
+
+        dna[
+            new_node_idx * self._length_per_region : (new_node_idx + 1) * self._length_per_region
+        ] = self._dna[
+            old_node_idx * self._length_per_region : (old_node_idx + 1) * self._length_per_region
+        ]
+
+        return dna
+
+    def _update_addable_nodes(
+        self,
+        node_dependencies: Dict[int, Set[int]],
+        used_node_indices: List[int],
+        old_node_idx: int,
+    ) -> Tuple[set, List[int]]:
+        """ Update the set of addable nodes,
+        by reevaluating which nodes haves no dependencies
+        and remove nodes which were already placed from the dependencies.
+        """
+        addable_nodes = self._get_addable_nodes(node_dependencies)
+        used_node_indices.append(old_node_idx)
+        addable_nodes = addable_nodes.difference(used_node_indices)
```

Comment:
logically speaking, the only thing that actually happens here is that `old_node_idx` is removed from the addable nodes, right? bc all other entries in used_node_indices were already removed before? if this is the case, maybe we can simplify the code to express this logic?

Reply:
No, because the implementation of `_get_addable_nodes` is returning all nodes with no dependencies. I think it makes sense to change that implementation to return only nodes which have no dependencies _and_ are not used yet

Reply:
and then replace this function